### PR TITLE
[stats] Migrate bayesian engine to gaussian effect model; unlock priors + CUPED

### PR DESCRIPTION
**Pull request by lukesonnet**
THIS PR IS DESCRIBED IN THE SUMMARY BELOW AND THE FOLLOWING TWO CHILDREN WHICH HAVE BEEN MERGED IN TO IT: #2407 #2409 

## Summary

Removes our gaussian mean and binomial proportion bayesian models and replaces them with the gaussian effect model (already in use by quantile metrics).

This internal document covers many of the details for what and why: https://www.notion.so/growthbook/Bayesian-Model-Transition-567387e78f4444ba9718fa9d955eab1b

Documentation for this model that will be merged eventually can be found here: https://www.notion.so/growthbook/Statistical-Details-ad44401a63ea402a9aef2944541e56ba

Also tweaks `risk` to not always be relative when coming back from the stats engine; it now depends on the difference type.

## Testing

Already confirmed differences with existing models in above notion doc, will link python notebook at a later date.

TODO:
- [x] Test prior rescaling for absolute differences
- [X] Add unit tests for posterior distributions (this is covered by the existing quantile test)

Follow-up PRs:
- #2407
- #2409

### CODE REVIEW

File: packages/stats/gbstats/gbstats.py [hunk 1]
```diff
@@ -231,40 +218,12 @@ def get_configured_test(
...
-                stat_a,
-                stat_b,
-                GaussianBayesianConfig(**base_config, inverse=metric.inverse),
-            )
-        else:
-            raise ValueError(COMPARE_PROPORTION_NON_PROPORTION_ERROR)
+        return GaussianEffectABTest(
+            stat_a,
+            stat_b,
+            GaussianEffectBayesianConfig(**base_config, inverse=metric.inverse),
```
Comment (lukebrawleysmith) [inline packages/stats/gbstats/gbstats.py:225]: I know GaussianEffectBayesianConfig permits prior mean specification.  Can you please confirm that customer will be able to specify prior mean in the UI?  In the PR it looked like they could specify only prior SD, but maybe I was missing something.

File: packages/stats/gbstats/bayesian/tests.py [hunk 2]
```diff
@@ -1,68 +1,40 @@
...
-@dataclass
-class GaussianBayesianConfig(BayesianConfig):
-    prior_a: GaussianPrior = field(default_factory=GaussianPrior)
-    prior_b: GaussianPrior = field(default_factory=GaussianPrior)
-    epsilon: float = 1e-4
-
-
-@dataclass
-class EffectBayesianConfig(BayesianConfig):
+class GaussianEffectBayesianConfig(BayesianConfig):
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/stats/gbstats/bayesian/tests.py:37]: SG. I can revert to EffectBayesianConfig. Good suggestion.

File: packages/stats/gbstats/bayesian/tests.py [hunk 3]
```diff
@@ -153,197 +118,33 @@ def scale_result(
...
-            ),
-            risk=risk,
-        )
-        if self.scaled:
-            result = self.scale_result(
-                result, self.traffic_proportion_b, self.phase_length_days
+        # rescale prior if needed
+        if self.relative and config.prior_type == "absolute":
+            self.prior_effect = GaussianPrior(
+                config.prior_effect.mean / self.stat_a.unadjusted_mean,
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/stats/gbstats/bayesian/tests.py:139]: There's lots of stuff assuming that it is positive, but yeah, we should future proof and use absolute for sure.

File: packages/stats/gbstats/bayesian/tests.py [hunk 4]
```diff
@@ -153,197 +118,33 @@ def scale_result(
...
-                stddev=std_diff,
-            ),
-            risk=risk,
-        )
-        if self.scaled:
-            result = self.scale_result(
-                result, self.traffic_proportion_b, self.phase_length_days
+        elif not self.relative and config.prior_type == "relative":
+            self.prior_effect = GaussianPrior(
+                config.prior_effect.mean * self.stat_a.unadjusted_mean,
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/stats/gbstats/bayesian/tests.py:145]: ty

File: packages/stats/gbstats/bayesian/tests.py [hunk 5]
```diff
@@ -357,50 +158,45 @@ def compute_result(self):
...
+        )
+        self.mean_diff = (
+            (
+                data_mean / data_variance
+                + self.prior_effect.mean / self.prior_effect.variance
+            )
+            / post_prec
+            if self.prior_effect.informative
+            else data_mean
         )
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/stats/gbstats/bayesian/tests.py:192]: I don't think this is true. The numerator in `mean_diff` will be affected by the prior variance and mean, right?

File: packages/stats/gbstats/bayesian/tests.py [hunk 6]
```diff
@@ -357,50 +158,45 @@ def compute_result(self):
...
-            frequentist_diff(mu_a, mu_b, False),
-            np.sqrt(
-                frequentist_variance(sd_a**2, mu_a, 1, sd_b**2, mu_b, 1, False)
-            ),
-        )
+        self.std_diff = np.sqrt(1 / post_prec)
+
         ctw = self.chance_to_win(self.mean_diff, self.std_diff)
         ci = self.credible_interval(
             self.mean_diff, self.std_diff, self.alpha, log=False
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/stats/gbstats/bayesian/tests.py:198]: Yes, will look into this, thanks.

File: packages/stats/gbstats/gbstats.py (hunk 1 above)
Comment (lukesonnet) [reply packages/stats/gbstats/gbstats.py:225]: Since updated as per our conversation.

**Review by lukebrawleysmith — APPROVED** [review]
lgtm

**Pull request merged**